apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,341 @@
+---
+title: 'Review: `torch.func` Contribution'
+published: September 22, 2023
+author: kshiteej-kalambarkar
+description: '`torch.func` (previously known as `functorch`) is a PyTorch module
+designed to offer JAX-like transforms. Within this module, various higher-order
+functions, such as `grad`, `vmap`, and `vjp` are made accessible. These
+transforms help users to easily compute gradients for the paramters of their model
+or write batch-size agnostic code. The beauty of these transformations lies in
+their ability to compose with one another. Thanks to this, the process of
+calculating per-sample gradients becomes the straightforward application of
+`vmap(grad(model))`.'
+---
+
+<base target="_blank" />
+
+`torch.func` (previously known as `functorch`) is a PyTorch module designed to
+offer JAX-like transforms. Within this module, various higher-order
+functions, such as `grad`, `vmap`, and `vjp` are made accessible. These
+transforms help users to easily compute gradients for the paramters of their model
+or write batch-size agnostic code. The beauty of these transformations lies in
+their ability to compose with one another. Thanks to this, the process of
+calculating per-sample gradients becomes the straightforward application of
+`vmap(grad(model))`.
+
+Here are a few of the tasks we had the opportunity to tackle:
+
+## Adding Batching Rule for `vmap`
+`vmap` is a transformation that accepts a function `func` that operaters on non-batched
+data. It then generates a new function capable of applying the
+given function to a batched input. When processing a batched input, an additional
+dimension, denoted by `in_dims`, is introduced to indicate which dimension to
+apply the function over. Conceptually, it emulates a `for` loop that iterates
+through all the data points and stacks the results. Importantly, it performs this
+operation efficiently by pushing the `for` loop into the PyTorch operations,
+effectively vectorizing the computation.
+
+Consider the following example:
+```python
+import torch
+
+# Written to handle only single sample.
+def my_simple_model(feature_vec, weight):
+    return torch.dot(feature_vec, weight).relu()
+
+batch_size = 4
+batched_inputs = torch.randn(batch_size, 3)
+weight = torch.randn(3)
+
+# For Loop version
+expected = []
+for input in batched_inputs:
+    expected.append(my_simple_model(input, weight))
+expected = torch.stack(expected)
+
+# Vmap
+# `in_dims` specifies the dimension that should be mapped over.
+# In this case, we map only over 0-dim of `batched_inputs`.
+actual = torch.vmap(my_simple_model, in_dims=(0, None))(batched_inputs, weight)
+
+# Verify that the results match.
+torch.testing.assert_close(expected, actual)
+```
+
+To support `vmap` for PyTorch operators, we need to specify the batching rule
+i.e. how to map the given function over a batched input. This is similar to how
+PyTorch internally specifies the rule for gradient computation for operators.
```

Comment:
Let's remove the gradient computation example, as it assumes that the user knows how autograd is implemented. You are explaining how to implement this transformation in the next line anyway.